README.md
# Groonga CloudSearch Console


NEWS.md
# News

## v1.2.0

* Validate index field options.
* Ask administrator credentials at the first time access.

## v1.1.0 (2012-11-29)

* Support index field operations:
  * Define index field
  * Delete index field
* Support SDF batch upload.

## v1.0.0 (2012-10-29)

* Initial release.


test/index.test.js
var assert = require('chai').assert;
var Browser = require('zombie');
var Target = require('./test-utils').Target;
var utils = require('./test-utils').utils;

suite('first time', function() {
  var target = new Target();
  setup(function(done) {
    target.setup(done)
  });
  teardown(function() {
    target.teardown()
  });

  test('should redirect to password configuration', function(done) {
    var browser = new Browser();

    browser
      .visit(target.rootURL)
      .then(function() {
        assert.equal(browser.location.pathname, '/admin/password');
      })
      .then(done, done);
  });

  test('config admin user', function(done) {
    var browser = new Browser();

    utils.withAdminConfigured(target, browser)
      .then(function() {
        assert.equal(browser.location.pathname, '/');
        assert.equal(browser.text('title'), 'Groonga CloudSearch Console');
      })
      .then(done, done);
  });
});


suite('password configured', function() {
  var target = new Target();
  setup(function(done) {
    target.setup(done)
  });
  teardown(function() {
    target.teardown()
  });

  test('GET / without credentials', function(done) {
    var browser = new Browser();
    utils.withAdminConfigured(target, browser)
      .then(function() {
        var browserWithoutPassword = new Browser();
        browserWithoutPassword
          .visit(target.rootURL)
          .fail(function(error) {
            assert.equal(browserWithoutPassword.statusCode, 401);
            assert.isNotNull(error);
            done();
          })
      });
  });

  test('GET / with wrong password', function(done) {
    var browser = new Browser();
    utils.withAdminConfigured(target, browser)
      .then(function() {
        var browserWithWrongPassword = new Browser();
        browserWithWrongPassword.authenticate().basic('username', 'wrong-password');
        browserWithWrongPassword
          .visit(target.rootURL)
          .fail(function(error) {
            assert.equal(browserWithWrongPassword.statusCode, 401);
            assert.isNotNull(error);
            done();
          });
      });
  });
});


test/test-utils.js
var spawn = require('child_process').spawn;
var mkdirp = require('mkdirp');
var rimraf = require('rimraf');
var fs = require('fs');
var path = require('path');
var existsSync = fs.existsSync || path.existsSync; // to support older nodes than 0.8
var Browser = require('zombie');

function runServer(path, options, callback) {
  var command = spawn(path, options);
  command.stdout.on('data', function(data) {
    if (data.toString().match(/listening/)) {
      callback();
    }
    process.stdout.write(data);
  });
  command.stderr.on('data', function(data) {
    process.stderr.write(data);
  });
  return command;
}

var Target = function(options) {
  this.gcsConsolePort = 3335;
  this.gcsPort = 3334;
  this.databaseDir = __dirname + '/../test/tmp/gcs';
  this.databasePath = this.databaseDir + '/gcs';
  this.gcsPath = __dirname + '/../node_modules/.bin/gcs';
  this.gcsConsolePath = __dirname + '/../bin/gcs-console';
  this.rootURL = 'http://localhost:' + this.gcsConsolePort + '/';
  this.gcsConsoleHome = this.databaseDir + '/gcs-console';
};

Target.prototype = {
  gcs: null,
  gcsConsole: null,
  username: 'username',
  password: 'password',
  setup: function(done) {
    var self = this;
    var gcsOptions = [
      '--database-path', self.databasePath,
      '--port', self.gcsPort
    ];
    var gcsConsoleOptions = [
      '--port', self.gcsConsolePort,
      '--endpoint', 'http://127.0.0.1.xip.io:' + self.gcsPort,
      '--gcs-console-home', self.gcsConsoleHome
    ];

    if (!existsSync(self.gcsPath)) {
      var error = new Error('gcs executable is not found at ' + self.gcsPath + '. You need to setup gcs to test with gcs-console. Run "npm install gcs" (for the latest release) or "npm install git://github.com/groonga/gcs.git" (for the development)');
      return done(error);
    }

    mkdirp.sync(self.databaseDir);

    self.gcs = runServer(self.gcsPath, gcsOptions, function() {
      process.on('exit', function() {
        self.teardown();
      });

      self.gcsConsole = runServer(
        self.gcsConsolePath,
        gcsConsoleOptions,
        function() {
          done();
        }
      )
    });
  },
  teardown: function() {
    if (this.gcs) {
      this.gcs.kill();
      rimraf.sync(this.databaseDir);
    }
    if (this.gcsConsole) {
      this.gcsConsole.kill();
    }
  },
  createBrowser: function() {
    var browser = new Browser();
    var self = this;

    browser.authenticate().basic(self.username, self.password);
    return browser
      .visit(self.rootURL + 'admin/password')
      .then(function() {
        browser.fill('username', self.username);
        browser.fill('password', self.password);
        return browser.pressButton('Save');
      });
  }
};

module.exports.Target = Target;


var utils = {
  withAdminConfigured: function(target, browser) {
    var username = 'username';
    var password = 'password';

    browser.authenticate().basic(username, password);
    return browser
    .visit(target.rootURL + 'admin/password')
    .then(function() {
      browser.fill('username', username);
      browser.fill('password', password);
      return browser.pressButton('Save')
    });
  }
};

module.exports.utils = utils;


test/index-fields.test.js
var assert = require('chai').assert;
var Browser = require('zombie');
var Target = require('./test-utils').Target;
var utils = require('./test-utils').utils;

suite('dashboard', function() {
  var target = new Target();
  setup(function(done) {
    target.setup(done)
  });
  teardown(function() {
    target.teardown()
  });

  test('Create domain, add an index field and delete the index field', function(done) {
    var browser = new Browser();

    utils.withAdminConfigured(target, browser)
      .then(function() {
        return browser.visit(target.rootURL);
      })
      .then(function() {
        return browser.clickLink('Create New Domain');
      })
      .then(function() {
        browser.fill('domain_name', 'test');
        return browser.pressButton('Create')
      })
      .then(function() {
        return browser.clickLink('Index Fields');
      })
      .then(function() {
        assert.equal(browser.location.pathname, "/domain/test/index_fields");
        assert.equal(browser.text('.alert-warn'), "No IndexField is defined");
      })
      .then(function() {
        browser.fill('name', 'test');
        return browser.pressButton('Create')
      })
      .then(function() {
        assert.equal(browser.location.pathname, "/domain/test/index_fields");
        assert.equal(browser.text(".alert"), "IndexField successfully created");
      })
      .then(function() {
        assert.equal(browser.text("h3"), "Delete 'test' Index Field");
        assert.equal(browser.text(".modal-body p"), "Are you sure?");
        return browser.pressButton('Delete')
      })
      .then(function() {
        assert.equal(browser.text(".alert-info"), "IndexField successfully deleted");
        assert.equal(browser.location.pathname, "/domain/test/index_fields");
        assert.equal(browser.text('.alert-warn'), "No IndexField is defined");
      })
      .then(done, done);
  });
});


app.js
var express = require('express')
  , http = require('http')
  , path = require('path');
var flash = require('connect-flash');
var bcrypt = require('bcrypt');

var Config = require('./lib/config').Config;
var routes = {
  domains: require('./routes/domains'),
  admin: require('./routes/admin')
};

function setupApplication(app) {
  var auth;

  var config = new Config(app.get('home'));
  app.set('config', config);

  auth = function(req, res, next) {
    var adminUsername = config.data.adminUsername;
    var adminPassword = config.data.adminPassword;
    if (adminUsername && adminPassword) {
      return express.basicAuth(function(username, password) {
        return username === adminUsername && bcrypt.compareSync(password, adminPassword);
      })(req, res, next);
    } else {
      return res.redirect('/admin/password');
    }
  };

  app.configure(function(){
    app.set('views', __dirname + '/views');
    app.set('view engine', 'jade');
    app.use(express.favicon());
    app.use(express.logger('dev'));
    app.use(express.bodyParser());
    app.use(express.methodOverride());
    app.use(express.cookieParser('keyboard cat'));
    app.use(express.session({ cookie: { maxAge: 60000 }}));
    app.use(flash());
    app.use((function() {
      return function(req, res, next) {
        res.locals.info = req.flash('info');
        next();
      }
    })());
    app.use(express.static(path.join(__dirname, 'public')));
    app.use(require('./lib/cloudsearch').cloudsearch(app.get('endpoint')));
    app.use(app.router);
  });

  app.configure('development', function(){
  });

  app.get('/', auth, routes.domains.index);
  app.get('/domain/:name', auth, routes.domains.show);
  app.get('/domain/:name/search', auth, routes.domains.search);
  app.get('/domain/:name/index_fields', auth, routes.domains.indexFields);
  app.get('/domain/:name/upload', auth, routes.domains.upload);
  app.post('/domain/:name/upload', auth, routes.domains.uploadPost);
  app.get('/domain_create', auth, routes.domains.create);
  app.post('/domain_create', auth, routes.domains.createPost);
  app.post('/domain/:name/index_fields', auth, routes.domains.createIndexField);
  app.delete('/domain/:name/index_fields/:indexFieldName', auth, routes.domains.deleteIndexField);
  app.delete('/domain/:name', auth, routes.domains.delete);

  app.get('/admin/password', routes.admin.password);
  app.post('/admin/password', routes.admin.passwordPost);
}

module.exports.setupApplication = setupApplication;


routes/admin.js
var bcrypt = require('bcrypt');

exports.password = function(req, res) {
  var config = req.app.get('config');
  if (config.data.adminUsername && config.data.adminPassword) {
    return res.redirect('/');
  }
  return res.render('admin-password');
};

exports.passwordPost = function(req, res) {
  var config = req.app.get('config');

  if (config.data.adminUsername && config.data.adminPassword) {
    return res.redirect('/');
  }

  var username = req.body.username;
  var password = req.body.password;

  if (username && password) {
    config.data.adminUsername = username;
    config.data.adminPassword = bcrypt.hashSync(password, 10);
    config.save();
    return res.redirect('/');
  } else {
    return res.render('admin-password', {
      error: new Error('Username and Password should not be empty')
    });
  }
};


routes/domains.js
var querystring = require('querystring');
var _ = require('underscore');
var http = require('http');
var awssum = require('awssum');
var DocumentService = awssum.load('amazon/cloudsearch').DocumentService;
var url = require('url');
var fs = require('fs');

exports.index = function(req, res) {
  res.render('index', {action: 'index'});
};

function domainToEndpoint(configurationEndpoint, domain, type) {
  var domainIdComponents = domain.DomainId.split('/');

  return 'http://' + type + '-' + domainIdComponents[1] + '-' + domainIdComponents[0] + '.' + configurationEndpoint.hostname + ':' + configurationEndpoint.port;
}

function countBytes(string) {
  string = encodeURIComponent(string);
  var escapedPartsMatcher = /\%[0-9a-f][0-9a-f]/gi;
  var escapedParts = string.match(escapedPartsMatcher);
  var notEscapedParts = string.replace(escapedPartsMatcher, '');
  return notEscapedParts.length + (escapedParts ? escapedParts.length : 0);
}

function createGcsDocumentService(req, domain) {
  var documentService = new DocumentService({
    domainName: domain.DomainName,
    domainId: domain.DomainId
  });
  var configurationEndpoint = url.parse(req.app.get('endpoint'));
  var endpoint = url.parse(
    domainToEndpoint(configurationEndpoint, domain, 'doc')
  );

  documentService.host = function() {
    return endpoint.hostname;
  };
  documentService.addExtras = function(options, args) {
    options.protocol = endpoint.protocol;
    options.port = endpoint.port;
  };
  documentService.addCommonOptions = function(options, args) {
    options.headers['content-type'] = 'application/json';
    options.headers['content-length'] = countBytes(JSON.stringify(args.Docs));
  };
  return documentService;
}

function withDomain(req, res, callback) {
  var domain = _.where(res.locals.domains, {DomainName: req.params.name})[0];
  domain.isSelected = true;

  req.domain = domain;
  req.documentService = createGcsDocumentService(req, domain);

  // get index fields
  req.cloudsearch.DescribeIndexFields({
    DomainName: domain.DomainName
  }, function(error, data) {
    if (error) {
      res.status(500);
      res.render('error', {error: new Error(error.Message)});
      return;
    }
    var indexFields = convertToArray(data.Body.DescribeIndexFieldsResponse.DescribeIndexFieldsResult.IndexFields.member);

    req.indexFields = indexFields;
    callback(req, res);
  });

  return domain;
}

function convertToArray(data) {
  // awssum returns undefined for 0 item, object itself for 1 item,
  // array for >1 items. Make all of them as a Array.
  if (!data) {
    return [];
  } else if (!(data instanceof Array)) {
    return [data];
  } else {
    return data;
  }
}

function errorToRender(error) {
  var errorToRender = null;
  try {
    errorToRender = new Error(error.Body.Response.Errors.Error.Message);
  } catch(e) {
    errorToRender = e;
  }
  return errorToRender;
}

exports.show = function(req, res) {
  withDomain(req, res, function(req, res) {
    res.render('domain-show', {
      action: 'domain_show',
      domain: req.domain
    });
  });
};

exports.indexFields = function(req, res) {
  withDomain(req, res, function(req, res) {
    res.render('domain-index-fields', {
      action: 'domain_index_fields',
      domain: req.domain,
      indexFields: req.indexFields
    });
  });
};

exports.search = function(req, res) {
  withDomain(req, res, function(req, res) {
    var query = req.query.query;
    var size = 5;
    var start = Number(req.query.start || 0);

    if (query === undefined) {
      var locals = {
        action: 'domain_search',
        domain: req.domain,
        query: null,
        requestURL: null,
        results: null
      };
      res.render('domain-search', locals);
      return;
    }

    var indexFieldNames = req.indexFields.map(function(indexField) {
      return indexField.Options.IndexFieldName;
    });

    var paramsForSearch = {
      q: query,
      size: size,
      start: start,
      'return-fields': indexFieldNames.join(',')
    };

    var configurationEndpoint = url.parse(req.app.get('endpoint'));
    var requestURL = domainToEndpoint(configurationEndpoint, req.domain, 'search') + '/2011-02-01/search?' + querystring.stringify(paramsForSearch);

    var buffer = '';
    var results = null;
    http.get(requestURL, function(searchResponse) {
      searchResponse.setEncoding('utf8');
      searchResponse.on('data', function (chunk) {
        buffer += chunk;
      });
      searchResponse.on('end', function() {
        var results = JSON.parse(buffer);
        var locals = {
          action: 'domain_search',
          domain: req.domain,
          query: query,
          requestURL: requestURL,
          results: results,
          start: start,
          nextLink: null,
          previousLink: null
        };

        if (results.hits) {
          if (results.hits.found > start + results.hits.hit.length) {
            var nextLinkParams = {
              query: query,
              start: start + results.hits.hit.length
            };
            locals.nextLink = '/domain/' +
              req.domain.DomainName +
              '/search?' +
              querystring.stringify(nextLinkParams);
          }

          if (start - size >= 0) {
            var previousLinkParams = {
              query: query,
              start: start - size
            };
            locals.previousLink = '/domain/' +
              req.domain.DomainName +
              '/search?' +
              querystring.stringify(previousLinkParams);
          }
        }

        res.render('domain-search', locals);
      });
    });
    // TODO handle errors
  });
};

exports.create = function(req, res) {
  res.render('domain-create', {
    action: "domain_create",
    domain: null,
    creatingDomainName: null
  });
};

exports.createPost = function(req, res) {
  var domainName = req.body.domain_name;
  req.cloudsearch.CreateDomain({
    DomainName: domainName
  }, function(error, data) {
    if (error) {
      if (error.Message) {
        res.status(500);
        res.render('error', {error: new Error(error.Message)});
        return;
      }

      res.render('domain-create', {
        action: "domain_create",
        domain: null,
        creatingDomainName: domainName,
        error: errorToRender(error)
      });
      return;
    }

    var domainCreated = data.Body.CreateDomainResponse.CreateDomainResult.DomainStatus;
    req.flash('info', 'Domain successfully created');
    res.redirect('/domain/' + domainCreated.DomainName);
  });
};

exports.delete = function(req, res) {
  req.cloudsearch.DeleteDomain({
    DomainName: req.params.name
  }, function(error, data) {
    if (error) {
      res.status(500);
      res.render('error', {error: new Error(error.Message)});
      return;
    }

    req.flash('info', 'Domain successfully deleted');
    res.redirect('/');
  });
};

function onToTrue(str) {
  if (str === 'on') {
    return 'true';
  } else {
    return 'false';
  }
}

exports.createIndexField = function(req, res) {
  withDomain(req, res, function(req, res) {
    var request = {
      DomainName: req.domain.DomainName,
      IndexField: {
        IndexFieldName: req.body.name,
        IndexFieldType: req.body.type
      }
    };

    switch (req.body.type) {
      case 'text':
        request.IndexField.TextOptions = {
          FacetEnabled: onToTrue(req.body.facet),
          ResultEnabled: onToTrue(req.body.result)
        };
        break;
      case 'literal':
        request.IndexField.LiteralOptions = {
          FacetEnabled: onToTrue(req.body.facet),
          ResultEnabled: onToTrue(req.body.result),
          SearchEnabled: onToTrue(req.body.search)
        };
        break;
      case 'uint':
        break;
    }

    var doneCallback = function(error, data) {
      if (error) {
        res.status(500);
        res.render('domain-index-fields', {
          error: errorToRender(error),
          action: 'domain_index_fields',
          domain: req.domain,
          indexFields: req.indexFields
        });
        return;
      }
      req.flash('info', 'IndexField successfully created');
      res.redirect('/domain/' + req.domain.DomainName + '/index_fields');
    };

    req.cloudsearch.DefineIndexField(request, doneCallback);
  });
};

exports.deleteIndexField = function(req, res) {
  withDomain(req, res, function(req, res) {
    req.cloudsearch.DeleteIndexField({
      DomainName: req.domain.DomainName,
      IndexFieldName: req.params.indexFieldName
    }, function(error, data) {
      if (error) {
        res.status(500);
        res.render('domain-index-fields', {
          error: errorToRender(error),
          action: 'domain_index_fields',
          domain: req.domain,
          indexFields: req.indexFields
        });
        return;
      }
      req.flash('info', 'IndexField successfully deleted');
      res.redirect('/domain/' + req.domain.DomainName + '/index_fields');
    });
  });
};

exports.upload = function(req, res) {
  withDomain(req, res, function(req, res) {
    res.render('domain-upload', {
      action: 'domain_upload',
      domain: req.domain
    });
  });
};

exports.uploadPost = function(req, res) {
  withDomain(req, res, function(req, res) {
    fs.readFile(req.files.batch.path, function(err, data) {
      try {
        var docs = JSON.parse(data);
      } catch(e) {
        res.status(400);
        res.render('domain-upload', {
          action: 'domain_upload',
          domain: req.domain,
          error: new Error('Failed to parse JSON')
        });
      }
      var options = { Docs: docs };

      req.documentService.DocumentsBatch(options, function(err, data) {
        if (err) {
          res.status(400);
          res.render('domain-upload', {
            action: 'domain_upload',
            domain: req.domain,
            error: errorToRender(err)
          });
        }

        req.flash('info', 'SDF Batch Uploaded');
        res.redirect('/domain/' + req.domain.DomainName + '/upload');
      });
    });
  });
};


lib/cloudsearch.js
var awssum = require('awssum');
var CloudSearch = awssum.load('amazon/cloudsearch').CloudSearch;
var url = require('url');
var util = require('util');

var cloudsearch = function(endpoint) {
  var endpointURL = url.parse(endpoint);

  CloudSearch.prototype.host = function() {
     return endpointURL.hostname;
  };

  var cs = new CloudSearch({
     accessKeyId: 'dummy-access-key-id',
     secretAccessKey: 'dummy-access-key',
  });

  cs.addExtras = function(options, args) {
     options.protocol = endpointURL.protocol.replace(/:$/, '');
     options.port = endpointURL.port;
  };

  return function(req, res, next) {
    req.cloudsearch = cs;
    res.locals.configurationEndpointURL = endpoint;
    cs.DescribeDomains(function(error, data) {
      if (error) {
        res.status(500);
        res.render('error', {error: new Error(error.Message)});
        console.error("awssum reported an error.\n" + util.inspect(error));
        return;
      }

      var domains = data.Body.DescribeDomainsResponse.DescribeDomainsResult.DomainStatusList.member;
      if (!domains) {
        domains = [];
      } else if (!(domains instanceof Array)) {
        domains = [domains];
      }
      res.locals.domains = domains;
      next();
    });
  };
};

module.exports.cloudsearch = cloudsearch;


public/js/domain-index-fields.js
jQuery(document).ready(function($) {
  function IndexFieldsOptionsViewModel() {
    var self = this;

    /* observers */
    this.chosenType = ko.observable();
    this.searchChecked = ko.observable();
    this.facetChecked = ko.observable();
    this.resultChecked = ko.observable();

    /* treat choosable options for each type */
    this.searchChoosable = ko.computed(function() {
      return this.chosenType() === 'literal';
    }, this);
    this.facetChoosable = ko.computed(function() {
      return this.chosenType() !== 'uint';
    }, this);
    this.resultChoosable = ko.computed(function() {
      return this.chosenType() !== 'uint';
    }, this);

    /* set initial options for each type */
    this.chosenType.subscribe(function(newType) {
      switch (newType) {
        case 'text':
          self.searchChecked(true);
          self.facetChecked(false);
          self.resultChecked(false);
          break;
        case 'literal':
          self.searchChecked(false);
          self.facetChecked(false);
          self.resultChecked(false);
          break;
        case 'uint':
          self.searchChecked(true);
          self.facetChecked(true);
          self.resultChecked(true);
          break;
      }
    });

    /* 'facet' and 'result' are exclusive except for uint */
    this.facetChecked.subscribe(function(facetChecked) {
      if (self.chosenType() === 'uint') {
        return;
      }
      if (facetChecked) {
        self.resultChecked(false);
      }
    });
    this.resultChecked.subscribe(function(resultChecked) {
      if (self.chosenType() === 'uint') {
        return;
      }
      if (resultChecked) {
        self.facetChecked(false);
      }
    });
  }

  ko.applyBindings(new IndexFieldsOptionsViewModel());
});


routes/index.js
var querystring = require('querystring');
var _ = require('underscore');
var http = require('http');
var awssum = require('awssum');
var DocumentService = awssum.load('amazon/cloudsearch').DocumentService;
var url = require('url');
var fs = require('fs');

exports.index = function(req, res) {
  res.render('index', {action: 'index'});
};

function domainToEndpoint(configurationEndpoint, domain, type) {
  var domainIdComponents = domain.DomainId.split('/');

  return 'http://' + type + '-' + domainIdComponents[1] + '-' + domainIdComponents[0] + '.' + configurationEndpoint.hostname + ':' + configurationEndpoint.port;
}

function countBytes(string) {
  string = encodeURIComponent(string);
  var escapedPartsMatcher = /\%[0-9a-f][0-9a-f]/gi;
  var escapedParts = string.match(escapedPartsMatcher);
  var notEscapedParts = string.replace(escapedPartsMatcher, '');
  return notEscapedParts.length + (escapedParts ? escapedParts.length : 0);
}

function createGcsDocumentService(req, domain) {
  var documentService = new DocumentService({
    domainName: domain.DomainName,
    domainId: domain.DomainId
  });
  var configurationEndpoint = url.parse(req.app.get('endpoint'));
  var endpoint = url.parse(
    domainToEndpoint(configurationEndpoint, domain, 'doc')
  );

  documentService.host = function() {
    return endpoint.hostname;
  };
  documentService.addExtras = function(options, args) {
    options.protocol = endpoint.protocol;
    options.port = endpoint.port;
  };
  documentService.addCommonOptions = function(options, args) {
    options.headers['content-type'] = 'application/json';
    options.headers['content-length'] = countBytes(JSON.stringify(args.Docs));
  };
  return documentService;
}

function withDomain(req, res, callback) {
  var domain = _.where(res.locals.domains, {DomainName: req.params.name})[0];
  domain.isSelected = true;

  req.domain = domain;
  req.documentService = createGcsDocumentService(req, domain);

  // get index fields
  req.cloudsearch.DescribeIndexFields({
    DomainName: domain.DomainName
  }, function(error, data) {
    if (error) {
      res.status(500);
      res.render('error', {error: new Error(error.Message)});
      return;
    }
    var indexFields = convertToArray(data.Body.DescribeIndexFieldsResponse.DescribeIndexFieldsResult.IndexFields.member);

    req.indexFields = indexFields;
    callback(req, res);
  });

  return domain;
}

function convertToArray(data) {
  // awssum returns undefined for 0 item, object itself for 1 item,
  // array for >1 items. Make all of them as a Array.
  if (!data) {
    return [];
  } else if (!(data instanceof Array)) {
    return [data];
  } else {
    return data;
  }
}

function errorToRender(error) {
  var errorToRender = null;
  try {
    errorToRender = new Error(error.Body.Response.Errors.Error.Message);
  } catch(e) {
    errorToRender = e;
  }
  return errorToRender;
}

exports.domain = function(req, res) {
  withDomain(req, res, function(req, res) {
    res.render('domain-show', {
      action: 'domain_show',
      domain: req.domain
    });
  });
};

exports.domainIndexFields = function(req, res) {
  withDomain(req, res, function(req, res) {
    res.render('domain-index-fields', {
      action: 'domain_index_fields',
      domain: req.domain,
      indexFields: req.indexFields
    });
  });
};

exports.domainSearch = function(req, res) {
  withDomain(req, res, function(req, res) {
    var query = req.query.query;
    var size = 5;
    var start = Number(req.query.start || 0);

    if (query === undefined) {
      var locals = {
        action: 'domain_search',
        domain: req.domain,
        query: null,
        requestURL: null,
        results: null
      };
      res.render('domain-search', locals);
      return;
    }

    var indexFieldNames = req.indexFields.map(function(indexField) {
      return indexField.Options.IndexFieldName;
    });

    var paramsForSearch = {
      q: query,
      size: size,
      start: start,
      'return-fields': indexFieldNames.join(',')
    };

    var configurationEndpoint = url.parse(req.app.get('endpoint'));
    var requestURL = domainToEndpoint(configurationEndpoint, req.domain, 'search') + '/2011-02-01/search?' + querystring.stringify(paramsForSearch);

    var buffer = '';
    var results = null;
    http.get(requestURL, function(searchResponse) {
      searchResponse.setEncoding('utf8');
      searchResponse.on('data', function (chunk) {
        buffer += chunk;
      });
      searchResponse.on('end', function() {
        var results = JSON.parse(buffer);
        var locals = {
          action: 'domain_search',
          domain: req.domain,
          query: query,
          requestURL: requestURL,
          results: results,
          start: start,
          nextLink: null,
          previousLink: null
        };

        if (results.hits) {
          if (results.hits.found > start + results.hits.hit.length) {
            var nextLinkParams = {
              query: query,
              start: start + results.hits.hit.length
            };
            locals.nextLink = '/domain/' +
              req.domain.DomainName +
              '/search?' +
              querystring.stringify(nextLinkParams);
          }

          if (start - size >= 0) {
            var previousLinkParams = {
              query: query,
              start: start - size
            };
            locals.previousLink = '/domain/' +
              req.domain.DomainName +
              '/search?' +
              querystring.stringify(previousLinkParams);
          }
        }

        res.render('domain-search', locals);
      });
    });
    // TODO handle errors
  });
};

exports.domainCreate = function(req, res) {
  res.render('domain-create', {
    action: "domain_create",
    domain: null,
    creatingDomainName: null
  });
};

exports.domainCreatePost = function(req, res) {
  var domainName = req.body.domain_name;
  req.cloudsearch.CreateDomain({
    DomainName: domainName
  }, function(error, data) {
    if (error) {
      if (error.Message) {
        res.status(500);
        res.render('error', {error: new Error(error.Message)});
        return;
      }

      res.render('domain-create', {
        action: "domain_create",
        domain: null,
        creatingDomainName: domainName,
        error: errorToRender(error)
      });
      return;
    }

    var domainCreated = data.Body.CreateDomainResponse.CreateDomainResult.DomainStatus;
    req.flash('info', 'Domain successfully created');
    res.redirect('/domain/' + domainCreated.DomainName);
  });
};

exports.domainDelete = function(req, res) {
  req.cloudsearch.DeleteDomain({
    DomainName: req.params.name
  }, function(error, data) {
    if (error) {
      res.status(500);
      res.render('error', {error: new Error(error.Message)});
      return;
    }

    req.flash('info', 'Domain successfully deleted');
    res.redirect('/');
  });
};

function onToTrue(str) {
  if (str === 'on') {
    return 'true';
  } else {
    return 'false';
  }
}

exports.domainCreateIndexField = function(req, res) {
  withDomain(req, res, function(req, res) {
    var request = {
      DomainName: req.domain.DomainName,
      IndexField: {
        IndexFieldName: req.body.name,
        IndexFieldType: req.body.type
      }
    };

    switch (req.body.type) {
      case 'text':
        request.IndexField.TextOptions = {
          FacetEnabled: onToTrue(req.body.facet),
          ResultEnabled: onToTrue(req.body.result)
        };
        break;
      case 'literal':
        request.IndexField.LiteralOptions = {
          FacetEnabled: onToTrue(req.body.facet),
          ResultEnabled: onToTrue(req.body.result),
          SearchEnabled: onToTrue(req.body.search)
        };
        break;
      case 'uint':
        break;
    }

    var doneCallback = function(error, data) {
      if (error) {
        res.status(500);
        res.render('domain-index-fields', {
          error: errorToRender(error),
          action: 'domain_index_fields',
          domain: req.domain,
          indexFields: req.indexFields
        });
        return;
      }
      req.flash('info', 'IndexField successfully created');
      res.redirect('/domain/' + req.domain.DomainName + '/index_fields');
    };

    req.cloudsearch.DefineIndexField(request, doneCallback);
  });
};

exports.domainDeleteIndexField = function(req, res) {
  withDomain(req, res, function(req, res) {
    req.cloudsearch.DeleteIndexField({
      DomainName: req.domain.DomainName,
      IndexFieldName: req.params.indexFieldName
    }, function(error, data) {
      if (error) {
        res.status(500);
        res.render('domain-index-fields', {
          error: errorToRender(error),
          action: 'domain_index_fields',
          domain: req.domain,
          indexFields: req.indexFields
        });
        return;
      }
      req.flash('info', 'IndexField successfully deleted');
      res.redirect('/domain/' + req.domain.DomainName + '/index_fields');
    });
  });
};

exports.domainUpload = function(req, res) {
  withDomain(req, res, function(req, res) {
    res.render('domain-upload', {
      action: 'domain_upload',
      domain: req.domain
    });
  });
};

exports.domainUploadPost = function(req, res) {
  withDomain(req, res, function(req, res) {
    fs.readFile(req.files.batch.path, function(err, data) {
      try {
        var docs = JSON.parse(data);
      } catch(e) {
        res.status(400);
        res.render('domain-upload', {
          action: 'domain_upload',
          domain: req.domain,
          error: new Error('Failed to parse JSON')
        });
      }
      var options = { Docs: docs };

      req.documentService.DocumentsBatch(options, function(err, data) {
        if (err) {
          res.status(400);
          res.render('domain-upload', {
            action: 'domain_upload',
            domain: req.domain,
            error: errorToRender(err)
          });
        }

        req.flash('info', 'SDF Batch Uploaded');
        res.redirect('/domain/' + req.domain.DomainName + '/upload');
      });
    });
  });
};

exports.adminPassword = function(req, res) {
  var config = req.app.get('config');
  if (config.data.adminUsername && config.data.adminPassword) {
    return res.redirect('/');
  }
  return res.render('admin-password');
};

var Config = require('../lib/config').Config;
exports.adminPasswordPost = function(req, res) {
  var config = req.app.get('config');

  if (config.data.adminUsername && config.data.adminPassword) {
    return res.redirect('/');
  }

  var username = req.body.username;
  var password = req.body.password;

  if (username && password) {
    config.data.adminUsername = username;
    config.data.adminPassword = password;
    config.save();
    return res.redirect('/');
  } else {
    return res.render('admin-password', {
      error: new Error('Username and Password should not be empty')
    });
  }
};
